File: src/main/java/ru/iqtranslit/rules/SchemeRuleSet.java
package ru.iqtranslit.rules;

import ru.iqtranslit.model.StandardScheme;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;

public final class SchemeRuleSet {
    private final StandardScheme scheme;
    private final Map<Character, String> strictMap;
    private final Map<Character, List<String>> extendedMap;

    public SchemeRuleSet(
            StandardScheme scheme,
            Map<Character, String> strictMap,
            Map<Character, List<String>> extendedMap
    ) {
        this.scheme = scheme;
        this.strictMap = strictMap;
        this.extendedMap = extendedMap;
    }

    public StandardScheme getScheme() {
        return scheme;
    }

    public String strictToken(char sourceChar) {
        String mapped = strictMap.get(Character.toLowerCase(sourceChar));
        if (mapped == null) {
            return String.valueOf(sourceChar);
        }
        return applyCase(mapped, sourceChar);
    }

    public String strictToken(String source, int index) {
        char sourceChar = source.charAt(index);
        if (scheme == StandardScheme.YANDEX_MAPS_STYLE) {
            return yandexStrictToken(source, index, sourceChar);
        }
        if (scheme == StandardScheme.BGN_PCGN_1947) {
            return bgnPcgnStrictToken(source, index, sourceChar);
        }
        if (scheme == StandardScheme.GOST_R_52290_2004) {
            return gost52290StrictToken(source, index, sourceChar);
        }
        return strictToken(sourceChar);
    }

    public List<String> extendedTokens(String source, int index) {
        char sourceChar = source.charAt(index);
        char lower = Character.toLowerCase(sourceChar);
        String strict = strictToken(source, index);
        if (strict.equals(String.valueOf(sourceChar)) && !strictMap.containsKey(lower)) {
            return Collections.singletonList(String.valueOf(sourceChar));
        }

        LinkedHashSet<String> options = new LinkedHashSet<String>();
        options.add(strict);

        List<String> mappedAlternatives = extendedMap.get(lower);
        if (mappedAlternatives != null) {
            for (String alternative : mappedAlternatives) {
                options.add(applyCase(alternative, sourceChar));
            }
        }

        addContextOptions(options, source, index, sourceChar, lower);
        return Collections.unmodifiableList(new ArrayList<String>(options));
    }

    private void addContextOptions(
            LinkedHashSet<String> options,
            String source,
            int index,
            char sourceChar,
            char lower
    ) {
        if (lower == 'е' && isWordStart(source, index)) {
            options.add(applyCase("ye", sourceChar));
            options.add(applyCase("je", sourceChar));
        }
        if (lower == 'ё') {
            options.add(applyCase("yo", sourceChar));
            options.add(applyCase("jo", sourceChar));
        }
        if (lower == 'й') {
            options.add(applyCase("y", sourceChar));
            options.add(applyCase("j", sourceChar));
            options.add(applyCase("i", sourceChar));
        }
        if (lower == 'х') {
            options.add(applyCase("kh", sourceChar));
            options.add(applyCase("h", sourceChar));
            options.add(applyCase("x", sourceChar));
        }
        if (lower == 'щ') {
            options.add(applyCase("shch", sourceChar));
            options.add(applyCase("sch", sourceChar));
            options.add(applyCase("shh", sourceChar));
        }
        if (lower == 'ъ' || lower == 'ь') {
            options.add("");
            options.add("'");
        }
        if (lower == 'ю') {
            options.add(applyCase("yu", sourceChar));
            options.add(applyCase("iu", sourceChar));
            options.add(applyCase("ju", sourceChar));
        }
        if (lower == 'я') {
            options.add(applyCase("ya", sourceChar));
            options.add(applyCase("ia", sourceChar));
            options.add(applyCase("ja", sourceChar));
        }
    }

    private static boolean isWordStart(String source, int index) {
        if (index == 0) {
            return true;
        }
        char prev = source.charAt(index - 1);
        return !Character.isLetter(prev);
    }

    private String yandexStrictToken(String source, int index, char sourceChar) {
        char lower = Character.toLowerCase(sourceChar);

        if (lower == 'е') {
            if (isWordStart(source, index) || followsYandexYeBoundary(source, index)) {
                return applyCase("ye", sourceChar);
            }
            return applyCase("e", sourceChar);
        }

        if (lower == 'щ') {
            return applyCase("sch", sourceChar);
        }

        if (lower == 'ъ') {
            return "";
        }

        if (lower == 'ь') {
            if (index + 1 < source.length() && Character.toLowerCase(source.charAt(index + 1)) == 'и') {
                return applyCase("y", sourceChar);
            }
            return "";
        }

        if ((lower == 'ы' || lower == 'и') && isYandexIyEnding(source, index)) {
            return applyCase("i", sourceChar);
        }

        return strictToken(sourceChar);
    }

    private String bgnPcgnStrictToken(String source, int index, char sourceChar) {
        char lower = Character.toLowerCase(sourceChar);

        if (lower == 'е') {
            if (isWordStart(source, index) || followsYeBoundary(source, index)) {
                return applyCase("ye", sourceChar);
            }
            return applyCase("e", sourceChar);
        }

        if (lower == 'ё') {
            if (isWordStart(source, index) || followsYeBoundary(source, index)) {
                return applyCase("yë", sourceChar);
            }
            return applyCase("ë", sourceChar);
        }

        return strictToken(sourceChar);
    }

    private String gost52290StrictToken(String source, int index, char sourceChar) {
        char lower = Character.toLowerCase(sourceChar);

        if (lower == 'е') {
            if (isWordStart(source, index) || followsYeBoundary(source, index)) {
                return applyCase("ye", sourceChar);
            }
            return applyCase("e", sourceChar);
        }

        if (lower == 'ё') {
            if (isWordStart(source, index) || followsYeBoundary(source, index)) {
                return applyCase("yo", sourceChar);
            }
            char prev = index > 0 ? Character.toLowerCase(source.charAt(index - 1)) : 0;
            if (prev == 'ч' || prev == 'ш' || prev == 'щ' || prev == 'ж') {
                return applyCase("e", sourceChar);
            }
            return applyCase("ye", sourceChar);
        }

        return strictToken(sourceChar);
    }

    private static boolean followsYandexYeBoundary(String source, int index) {
        if (index == 0) {
            return true;
        }
        char prev = Character.toLowerCase(source.charAt(index - 1));
        return isVowel(prev) || prev == 'ь' || prev == 'ъ';
    }

    private static boolean followsYeBoundary(String source, int index) {
        if (index == 0) {
            return true;
        }
        char prev = Character.toLowerCase(source.charAt(index - 1));
        return isVowel(prev) || prev == 'ь' || prev == 'ъ' || prev == 'й';
    }

    private static boolean isYandexIyEnding(String source, int index) {
        if (index + 1 >= source.length()) {
            return false;
        }
        char next = Character.toLowerCase(source.charAt(index + 1));
        if (next != 'й') {
            return false;
        }
        return index + 2 == source.length() || !Character.isLetter(source.charAt(index + 2));
    }

    private static boolean isVowel(char value) {
        return value == 'а'
                || value == 'е'
                || value == 'ё'
                || value == 'и'
                || value == 'о'
                || value == 'у'
                || value == 'ы'
                || value == 'э'
                || value == 'ю'
                || value == 'я';
    }

    private static String applyCase(String token, char sourceChar) {
        if (token.isEmpty()) {
            return token;
        }
        if (!Character.isUpperCase(sourceChar)) {
            return token;
        }
        if (token.length() == 1) {
            return token.toUpperCase();
        }
        return Character.toUpperCase(token.charAt(0)) + token.substring(1);
    }
}

File: python/src/iqtranslit_py/scheme_ruleset.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List

from .models import StandardScheme


def _ordered_add(target: Dict[str, None], value: str) -> None:
    target.setdefault(value, None)


@dataclass(frozen=True)
class SchemeRuleSet:
    scheme: StandardScheme
    strict_map: Dict[str, str]
    extended_map: Dict[str, List[str]]

    def strict_token(self, source_char: str) -> str:
        mapped = self.strict_map.get(source_char.lower())
        if mapped is None:
            return source_char
        return self._apply_case(mapped, source_char)

    def strict_token_at(self, source: str, index: int) -> str:
        source_char = source[index]
        if self.scheme == StandardScheme.YANDEX_MAPS_STYLE:
            return self._yandex_strict_token(source, index, source_char)
        if self.scheme == StandardScheme.BGN_PCGN_1947:
            return self._bgn_pcgn_strict_token(source, index, source_char)
        if self.scheme == StandardScheme.GOST_R_52290_2004:
            return self._gost_52290_strict_token(source, index, source_char)
        return self.strict_token(source_char)

    def extended_tokens(self, source: str, index: int) -> List[str]:
        source_char = source[index]
        lower = source_char.lower()
        strict = self.strict_token_at(source, index)
        if strict == source_char and lower not in self.strict_map:
            return [source_char]

        options: Dict[str, None] = {}
        _ordered_add(options, strict)

        mapped_alternatives = self.extended_map.get(lower)
        if mapped_alternatives is not None:
            for alternative in mapped_alternatives:
                _ordered_add(options, self._apply_case(alternative, source_char))

        self._add_context_options(options, source, index, source_char, lower)
        return list(options.keys())

    def _add_context_options(
        self,
        options: Dict[str, None],
        source: str,
        index: int,
        source_char: str,
        lower: str,
    ) -> None:
        if lower == "е" and self._is_word_start(source, index):
            _ordered_add(options, self._apply_case("ye", source_char))
            _ordered_add(options, self._apply_case("je", source_char))
        if lower == "ё":
            _ordered_add(options, self._apply_case("yo", source_char))
            _ordered_add(options, self._apply_case("jo", source_char))
        if lower == "й":
            _ordered_add(options, self._apply_case("y", source_char))
            _ordered_add(options, self._apply_case("j", source_char))
            _ordered_add(options, self._apply_case("i", source_char))
        if lower == "х":
            _ordered_add(options, self._apply_case("kh", source_char))
            _ordered_add(options, self._apply_case("h", source_char))
            _ordered_add(options, self._apply_case("x", source_char))
        if lower == "щ":
            _ordered_add(options, self._apply_case("shch", source_char))
            _ordered_add(options, self._apply_case("sch", source_char))
            _ordered_add(options, self._apply_case("shh", source_char))
        if lower == "ъ" or lower == "ь":
            _ordered_add(options, "")
            _ordered_add(options, "'")
        if lower == "ю":
            _ordered_add(options, self._apply_case("yu", source_char))
            _ordered_add(options, self._apply_case("iu", source_char))
            _ordered_add(options, self._apply_case("ju", source_char))
        if lower == "я":
            _ordered_add(options, self._apply_case("ya", source_char))
            _ordered_add(options, self._apply_case("ia", source_char))
            _ordered_add(options, self._apply_case("ja", source_char))

    @staticmethod
    def _is_word_start(source: str, index: int) -> bool:
        if index == 0:
            return True
        prev = source[index - 1]
        return not prev.isalpha()

    def _yandex_strict_token(self, source: str, index: int, source_char: str) -> str:
        lower = source_char.lower()
        if lower == "е":
            if self._is_word_start(source, index) or self._follows_yandex_ye_boundary(source, index):
                return self._apply_case("ye", source_char)
            return self._apply_case("e", source_char)
        if lower == "щ":
            return self._apply_case("sch", source_char)
        if lower == "ъ":
            return ""
        if lower == "ь":
            if index + 1 < len(source) and source[index + 1].lower() == "и":
                return self._apply_case("y", source_char)
            return ""
        if (lower == "ы" or lower == "и") and self._is_yandex_iy_ending(source, index):
            return self._apply_case("i", source_char)
        return self.strict_token(source_char)

    def _bgn_pcgn_strict_token(self, source: str, index: int, source_char: str) -> str:
        lower = source_char.lower()
        if lower == "е":
            if self._is_word_start(source, index) or self._follows_ye_boundary(source, index):
                return self._apply_case("ye", source_char)
            return self._apply_case("e", source_char)
        if lower == "ё":
            if self._is_word_start(source, index) or self._follows_ye_boundary(source, index):
                return self._apply_case("yë", source_char)
            return self._apply_case("ë", source_char)
        return self.strict_token(source_char)

    def _gost_52290_strict_token(self, source: str, index: int, source_char: str) -> str:
        lower = source_char.lower()
        if lower == "е":
            if self._is_word_start(source, index) or self._follows_ye_boundary(source, index):
                return self._apply_case("ye", source_char)
            return self._apply_case("e", source_char)
        if lower == "ё":
            if self._is_word_start(source, index) or self._follows_ye_boundary(source, index):
                return self._apply_case("yo", source_char)
            prev = source[index - 1].lower() if index > 0 else ""
            if prev in ("ч", "ш", "щ", "ж"):
                return self._apply_case("e", source_char)
            return self._apply_case("ye", source_char)
        return self.strict_token(source_char)

    @staticmethod
    def _follows_yandex_ye_boundary(source: str, index: int) -> bool:
        if index == 0:
            return True
        prev = source[index - 1].lower()
        return SchemeRuleSet._is_vowel(prev) or prev == "ь" or prev == "ъ"

    @staticmethod
    def _follows_ye_boundary(source: str, index: int) -> bool:
        if index == 0:
            return True
        prev = source[index - 1].lower()
        return SchemeRuleSet._is_vowel(prev) or prev in ("ь", "ъ", "й")

    @staticmethod
    def _is_yandex_iy_ending(source: str, index: int) -> bool:
        if index + 1 >= len(source):
            return False
        next_char = source[index + 1].lower()
        if next_char != "й":
            return False
        return index + 2 == len(source) or not source[index + 2].isalpha()

    @staticmethod
    def _is_vowel(value: str) -> bool:
        return value in ("а", "е", "ё", "и", "о", "у", "ы", "э", "ю", "я")

    @staticmethod
    def _apply_case(token: str, source_char: str) -> str:
        if token == "":
            return token
        if not source_char.isupper():
            return token
        if len(token) == 1:
            return token.upper()
        return token[0].upper() + token[1:]
